### PR TITLE
JAMES-3516 Implement ThreadIdGuessingAlgorithm for Distributed module

### PR DESCRIPTION
**Pull request by quantranhong1999**
The problem now is that the new message can't find related thread data from the thread table.
My guess is the Cassandra write is not finished yet before other new messages coming, or maybe even my code writing to Cassandra is not working.
I am working on this.

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import org.apache.james.mailbox.model.ThreadId;
+import org.apache.james.mailbox.store.mail.ThreadIdGuessingAlgorithm;
+import org.apache.james.mailbox.store.mail.model.MimeMessageId;
+import org.apache.james.mailbox.store.mail.model.Subject;
+import org.apache.james.mailbox.store.search.SearchUtil;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class StoreThreadIdGuessingAlgorithm implements ThreadIdGuessingAlgorithm {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java:46]: Store => Cassandra ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public Mono<ThreadId> guessThreadIdReactive(MessageId messageId, Optional<MimeMessageId> mimeMessageId, Optional<MimeMessageId> inReplyTo, Optional<List<MimeMessageId>> references, Optional<Subject> subject, MailboxSession session) {
+        Set<MimeMessageId> mimeMessageIds = buildMimeMessageIdSet(mimeMessageId, inReplyTo, references);
+        Optional<Subject> baseSubject = subject.map(value -> new Subject(SearchUtil.getBaseSubject(value.getValue())));
+        return Flux.from(threadDAO.selectSome(session.getUser(), mimeMessageIds))
+            .filter(pair -> pair.getLeft().equals(baseSubject))
+            .next()
+            .map(Pair::getRight)
+            .switchIfEmpty(Mono.just(ThreadId.fromBaseMessageId(messageId)))
+            .doOnSuccess(threadId -> threadDAO.insertSome(session.getUser(), mimeMessageIds, messageId, threadId, baseSubject));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java:65]: I guess you need to subscribe (and schedule on an elastic scheduler) the insertSome mono

Comment (quantranhong1999) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java:65]: I tried .blockLast
image
And then test guessing threadId for two newly related messages (non message before):
image
The test was green. It's look like guessThreadId using Cassandra is ok now. 

But the getThread API using ElasticSearch do not work properly. Even I tried to plug SearchThreadIdGuessingAlgorithm to JMAP level module and the result using ES is not proper.
The inmemory test using ScanningSearch is ok, so my guess problem is at ES. I will continue working on this.

Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/StoreThreadIdGuessingAlgorithm.java:65]: `doOnSuccess(id -> xxx.subscribeOn(Scheduler.elastic()).subscribe())`

**Comment by quantranhong1999** [comment]
At DistributedThreadGetTest:
- I added await for ES to ensure it can index all-new message
- I debugged a test that adds new related messages: they have the same threadId => So guessThreadId API works properly now.
But then get that thread do not work properly. The thread with 3 message id just returns 1 message id.

At CassandraThreadIdGuessingAlgorithmTest which uses SimpleMessageSearchIndex does not work properly also.

This is really weird. My guess is we have a bug when SearchIndex indexes threadId which wraps CassandraMessageId (because it still works properly with InMemoryMessageId).

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMetadata.java [hunk 3]
```diff
@@ -286,7 +286,7 @@ public static CassandraMessageMetadata from(MailboxMessage mailboxMessage, BlobI
                 .composedMessageId(new ComposedMessageId(mailboxMessage.getMailboxId(), mailboxMessage.getMessageId(), mailboxMessage.getUid()))
                 .modSeq(mailboxMessage.getModSeq())
                 .flags(mailboxMessage.createFlags())
-                .threadId(ThreadId.fromBaseMessageId(mailboxMessage.getMessageId()))
+                .threadId(mailboxMessage.getThreadId())
```
Comment (quantranhong1999) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMetadata.java:289]: I found it :)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 4]
```diff
@@ -165,6 +165,7 @@ trait ThreadGetContract {
         MessageManager.AppendCommand.from(Message.Builder.of.setSubject("Test")
         .setMessageId("Message-ID-1")
           .setBody("testmail", StandardCharsets.UTF_8)))
+    Thread.sleep(1000)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:168]: I prefer to not use the .sleep and wrap the assertion in an await.

`Await.untilAsserted(() -> ...)`

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 5]
```diff
@@ -31,16 +34,24 @@ import org.apache.james.jmap.core.ResponseObject.SESSION_STATE
...
 import org.apache.james.mime4j.stream.RawField
 import org.apache.james.modules.MailboxProbeImpl
 import org.apache.james.utils.DataProbeImpl
 import org.junit.jupiter.api.{BeforeEach, Test}
 
 trait ThreadGetContract {
+
+  protected def awaitMessageCount(mailboxIds: util.List[MailboxId], query: SearchQuery, messageCount: Long): Unit
+
+  @throws(classOf[IOException])
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:48]:
```suggestion
  @throws(classOf[IOException])
```

We IMO do not need this...

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 6]
```diff
@@ -193,7 +208,7 @@ trait ThreadGetContract {
       .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
       .body(request)
     .when
-      .post
+      .post.prettyPeek()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:211]: debug spotted

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 7]
```diff
@@ -273,7 +292,7 @@ trait ThreadGetContract {
...
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:295]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 8]
```diff
@@ -354,7 +349,7 @@ trait ThreadGetContract {
...
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:352]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 9]
```diff
@@ -436,7 +433,7 @@ trait ThreadGetContract {
...
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:436]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala [hunk 10]
```diff
@@ -515,7 +514,7 @@ trait ThreadGetContract {
       .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
       .body(request)
       .when
-      .post
+      .post.prettyPeek()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ThreadGetContract.scala:517]: idem

**Review by chibenwa — COMMENTED** [review]
Looks great to me ;-)

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]
Can you squash @quantranhong1999  ?

**Comment by quantranhong1999** [comment]
> Can you squash @quantranhong1999 ?

Done. We need this to be merged https://github.com/apache/james-project/pull/550. Then I can remove the NO REVIEW commit.

**Comment by quantranhong1999** [comment]
The CI build failed with weird errors. Can someone restart it?

**Comment by Arsnael** [comment]
yes maybe a network issue with jib. I restarted the build

**Comment by quantranhong1999** [comment]
```
org.apache.james.backends.es.v7.ElasticSearchIndexerTest.deleteMessagesShouldNotThrowWhenEmptyList

Failing for the past 1 build (Since Failed#8 )
Took 0 ms.
Error Message
Could not initialize class org.apache.james.backends.es.v7.DockerElasticSearchSingleton
Stacktrace
java.lang.NoClassDefFoundError: Could not initialize class org.apache.james.backends.es.v7.DockerElasticSearchSingleton
	at org.apache.james.backends.es.v7.ElasticSearchIndexerTest.<init>(ElasticSearchIndexerTest.java:61)
```
These seem not related, green on my local btw.

**Comment by Arsnael** [comment]
Looks like it's not... restarted

**Pull request merged**